RFC: TFX Generic Container Component Execution (#156)
tensorflow/community: rfcs/20190829-tfx-container-component-execution.md

# TFX Container Component Execution Proposal

Status         | Proposed
:------------- | :-------
**Author(s)**  | Ajay Gopinathan (<a href="mailto:ajaygopinathan@google.com">ajaygopinathan@google.com</a>), Hongye Sun (<a href="mailto:hongyes@google.com">hongyes@google.com</a>), Makoto Uchida (<a href="mailto:muchida@google.com">muchida@google.com</a>), Ruoyu Liu (<a href="mailto:ruoyu@google.com">ruoyu@google.com</a>)
**Sponsor(s)** | Konstantinos Katsiapis (katsiapis@google.com), Pavel Dournov (dournov@google.com), Ramesh Chandra (rameshuc@google.com)
**Updated**    | 2019-08-29

> ## Objective

Reviewer: Operations within the container may be complex and could benefit from the ml_metadata module.  Can there be a way for the various steps in a container to "log" their events similar to the way components do that now?....In that there wouldn't just be one "component" in the container, but may be multiple operations; each of which would be recorded in the same ml_metadata that the rest of the pipeline uses.

This RFC proposes an orchestrator agnostic way to reliably execute a user’s
container in the TFX pipeline. The proposal can support:

*   Running an arbitrary container in either local docker environment or remote

>     k8s cluster.

Reviewer 2: > k8s
Kubernetes

*   Passing data into the container
*   Passing output data from the container
*   Capturing logs from the container
*   Handling errors and retries

> *   Cancelling the container execution if the pipeline is terminated

Reviewer: Could you elaborate more on what "Cancelling" means including a timeline of what happens later in this document?  For instance, are their provisions to signal the running container so that it can perform a clean shutdown?

Reviewer: Container Lifetime:  Container instances may benefit from internal caching or reductions in startup time (ie applications within container taking some time to start up).  Ideally, the container instance could stay alive across multiple pipeline calls of it.

## Motivation

Currently, in a TFX pipeline, there is no way to execute a generic container as

> one of its steps. Without this feature, users cannot bring their own containers

Reviewer 2: > Currently, in a TFX pipeline, there is no way to execute a generic container as
one of its steps.
Currently, the execution of a generic container as a step in a TFX pipeline is not supported.

> into the pipeline. This blocks following use cases:

Reviewer 2: > This blocks *the* following use cases

The execution may occurs in local docker or in a remote kubernetes cluster.

### Existing solutions

#### Kubeflow Pipeline (KFP) ContainerOp

Today, KFP’s ContainerOp leverages
[Argo container template API](https://github.com/argoproj/argo/blob/master/pkg/apis/workflow/v1alpha1/workflow_types.go)
to launch user’s container in a k8s pod. Argo as the orchestrator controls when
to launch the POD and it uses a sidecar container to report output files back
and wait for user’s container to complete. We are not proposing to use Argo API
because of the following reasons:

*   Argo’s API is orchestrator-specific and cannot be ported to Airflow or local
    runners.
*   Argo’s API doesn’t provide an extensible way to run custom code before and
    after POD API which is critical to support metadata tracking and caching
    features.
*   Argo doesn’t provide an easy way to recover from user’s transient errors,
    which is critical in production workload.

#### Airflow k8s pod operator

Airflow supports to launch a k8s pod by an
[operator](https://github.com/apache/airflow/blob/master/airflow/contrib/operators/kubernetes_pod_operator.py).

> This approach is closer to what we are proposing in the document. However, we

Reviewer: Compared Existing Solutions is there any attempt to make a full replacement of one or more of those solutions (ie creating a better system versus creating a different system suited to TFX's needs)?  This would not be a requirement but a nice to have.

cannot directly use the operator because:

*   Airflow operator requires to be run inside an Airflow pipeline which is not
    the case for local and KF runners.
*   Airflow operator exposes a subset of POD’s API, where we want to expose the
    full pod spec to the user.
*   Airflow operator doesn’t provide a reliable way to retry user’s container
    and recover from transient errors.
*   Airflow does not support initializing operator inside another operator.
    Going back to use multiple Airflow operators for a component is a regression
    now that we have `BaseComponentLauncher` ready.

## Proposed Design

### TLDR

> We propose to solve the above problems by the following design.

Reviewer 2: > We propose to solve the above problems *with* the following design:

> *   Define container as an executor spec.

Reviewer 2: > Define _a_ container as an executor spec.

> *   Launch container by component launcher in either local docker or k8s pod.

Reviewer 2: > Launch _a_ container _via a_ component launcher in either _a_ local Docker or k8s pod.

> *   Use platform config to specify platform specific settings like k8s pod

Reviewer 2: > Use _a_ platform config to specify _a platform-specific_ settings config.

    config.

The proposed solution has following parts:

*   Extensible `ExecutorSpec` concept which can support container as an
    executor.
*   Extensible `BaseComponentLauncher` concept to support pluggable component
    launchers in tfx runner.
    *   `DockerComponentLauncher` which launches `ExecutorContainerSpec` in

>         docker environment.

Reviewer: Although initially targeted for "Docker", I would suggest changing this to something like "Any OCI container system, which includes Docker".  In other words, have a goal to work with any OCI container technology.

Author: sg. The spec is not specific to docker. We just provide an implementation for docker.

    *   `KubernetesPodComponentLauncher` which launches `ExecutorContainerSpec`
        in k8s environment.
*   Extensible `PlatformConfig` framework.
    *   `KubernetesPodPlatformConfig` to support k8s pod spec as a config.
    *   `DockerPlatformConfig` to support docker run configs.

### Architecture

Architecture that allows local container execution.

![TFX local container execution](20190829-tfx-container-component-execution/tfx-local-container-execution.png)

Architecture that allows Kubernetes container execution.

![TFX k8s container execution](20190829-tfx-container-component-execution/tfx-k8s-container-execution.png)

Class diagram that allows container execution

![TFX container execution_classes](20190829-tfx-container-component-execution/tfx-container-execution-classes.png)

### Python DSL experience

In order to use container base component in TFX DSL, user needs to do following
steps. Step 1 and Step 2 follow the DSL extension proposed by the other RFC
(https://github.com/tensorflow/community/pull/146).

> #### Step 1: Defines container based component by `ExecutorContainerSpec`

Reviewer 2: I think it means _with an_

```python
class MyContainerBasedExampleGen(BaseComponent):

  SPEC_CLASS = types.make_spec_class(
    inputs={
      "raw_input": ChannelParameter(type=standard_artifacts.ExternalArtifact),
    }
    outputs={
      "examples": ChannelParameter(type=standard_artifacts.Examples),
    }
    parameters={
      "num_csv_columns": ExecutionParameter(type=int),
    }
  )

>   EXECUTOR_SPEC = ExecutorContainerSpec(

Reviewer: (similar to previous comment):  Is ExecutorContainerSpec() specific to Docker?  Or, will it be correct for any OCI-compatible container solution?  If specific to Docker, rename it.

Author: Yes, It's not specific to Docker.

    container_image='gcr.io/my_project/my_example_gen:stable',
    command=['python'],
    args=['my_example_gen.py',
          '--input_csv_file', '{{ inputs.raw_input.uri }}',
          '--output_examples', '{{ outputs.examples.uri }}',
          '--num_csv_columns', '{{ exec_props.num_csv_columns }}' ],
  )
```

#### Step 2: Creates pipeline from container based component

```python
def create_pipeline():
  my_csv_file = Channel('CSVFile', uri="/path/to/csv_file")

  my_example_gen = MyContainerBasedExampleGen(
        raw_input=my_csv_file, num_csv_columns=20)

  return pipeline.Pipeline(
    pipeline_name = 'my_container_based_pipeline',
    pipeline_root = 'gs://path/to/root',
    components = [my_example_gen],
    ...
  )
```

> #### Step 3(a): Sets docker config via runner’s config

Reviewer 2: Want to continue to capitalize _Docker_ here as well for consistency?

```python
_ = BeamRunner(platform_configs={
  'MyContainerBasedExampleGen': [DockerPlatformConfig(volumes={...})]
}).run(create_pipeline())
```

> #### Step 3(b): Sets k8s platform config via runner’s config

Reviewer 3: ```suggestion
#### Step 3(b): Set k8s platform config via runner’s config
```

```python
_ = KubeflowDagRunner(platform_configs={
  'default': [KubernetesPodPlatformConfig(Pod().use_gcp_secret().spec()]
  'MyContainerBasedExampleGen': [
      KubernetesPodPlatformConfig(Pod(cpu=2, memory='1GB').spec())]}
).run(create_pipeline())
```

### Component launcher

> A component launcher launches a component by invoking driver, executor and

Reviewer 3: ```suggestion
A component launcher launches a component by invoking a driver, an executor and
```

> publisher. It understands how to launch a component executor from an

Reviewer 3: ```suggestion
a publisher. It understands how to launch a component executor from an
```

`ExecutorSpec`. The `BaseComponentLauncher` is an abstract base class with two
abstract methods:

*   `can_launch`: public method to check whether the launcher can launch an
    instance of `ExecutorSpec` with a specified `PlatformConfig` instance. The
    method will be used by `TfxRunner` to choose launcher for a component.
*   `_run_executor`: a protected method to launch an `ExecutorSpec` instance.
    The method is invoked by `BaseComponentLauncher.launch` method.

Subclasses of the base component launcher can support launching executors in
different target platforms. For example:

> *   `InProcessComponentLauncher` can launch an executor class in the same python

Reviewer 3: ```suggestion
*   `InProcessComponentLauncher` can launch an executor class in the same Python
```

    process.

> *   `DockerComponentLauncher` can launch a container executor in docker

Reviewer 3: ```suggestion
*   `DockerComponentLauncher` can launch a container executor in a Docker
```

    environment.

> *   `KubernetesPodComponentLauncher` can launch a container executor in k8s

Reviewer 3: ```suggestion
*   `KubernetesPodComponentLauncher` can launch a container executor in a k8s
```

    environment.

> *   A Dataflow launcher can launch a beam executor in Dataflow service.

Reviewer 2: > a beam executor
_Beam_

> *   Etc.

Reviewer 3: ```suggestion
```

      return false

    if not platform_spec:
      return true

    return isinstance(platform_spec, DockerPlatformConfig):

  def _run_executor(execution_id: int,
                   input_dict: Dict[Text, List[types.Artifact]],
                   output_dict: Dict[Text, List[types.Artifact]],
                   exec_properties: Dict[Text, Any]) -> None:
    # Docker launcher implementation
    ...

class KubernetesPodComponentLauncher(BaseComponentLauncher):
  @classmethod
  def can_launch(cls, executor_spec: ExecutorSpec,
                 platform_spec: Optional[PlatformConfig]) -> bool:
    if not isinstance(executor_spec, ExecutorContainerSpec):
      return false

    if not platform_spec:
      return true

    return isinstance(platform_spec, DockerPlatformConfig):

  def _run_executor(execution_id: int,
                   input_dict: Dict[Text, List[types.Artifact]],
                   output_dict: Dict[Text, List[types.Artifact]],
                   exec_properties: Dict[Text, Any]) -> None:
    # k8s pod launcher implementation
    …
```

### Platform config

Platform config carries platform specific configs. Usually, one platform config

> type maps to one type of component launcher type. For example,

Reviewer 3: ```suggestion
type maps to one component launcher type. For example,
```

instead of replacing them entirely. In this way, the patch layer doesn’t have to
specify the full content of a list or map.

The base pod spec layer is created from user’s container spec. The pod spec
includes a main container spec with image path and entrypoint of the container.

Default and component platform configs are configured by runner’s constructor.

For example:

```yaml
# base pod spec
apiVersion: v1
kind: Pod
spec:
  containers:
  - name: main
    image: tensorflow/tensorflow:v1.13
    command: ["python", "-c", "ml/app.py"]

# pipeline pod spec
spec:
  serviceAccountName: PipelineRunner
  containers:
  - name: main
    resources:
      limits:
        memory: "128Mi"
        cpu: "500m"

# component config pod spec
spec:
  containers:
  - name: main
    env:
    - name: MYSQL_ROOT_PASSWORD
      value: "password"

# final pod spec
apiVersion: v1
kind: Pod
spec:
  serviceAccountName: PipelineRunner
  containers:
  - name: main
    image: tensorflow/tensorflow:v1.13
    command: ["python", "-c", "ml/app.py"]
    resources:
      limits:
        memory: "128Mi"
        cpu: "500m"
    env:
    - name: MYSQL_ROOT_PASSWORD
      value: "password"
```

### TFX runner

> A `TFXRunner` compiles a logical pipeline into underlying orchestrator’s DSL. In

Reviewer 3: ```suggestion
A `TFXRunner` compiles a logical pipeline into the underlying orchestrator’s DSL. In
```

  @abc.abstractmethod
  def _run(self, pipeline, component_launcher_infos) -> Optional[Any]:
    pass
```

### Output interface

User container can receive a
[tmp directory path from default artifact store](https://github.com/tensorflow/community/blob/2c0b009ef955975b15a3cc18b1378e0ed38f394e/rfcs/20190904-tfx-generic-container-based-component.md#artifact-properties-after-execution-is-complete)
to write output data. The directory parameter will be called
`exec_properties.tmp_path`, which can be passed in as a command line argument.
The executor will look for `output.json` file under `exec_properties.tmp_path`
to get the outputs from the component. The output file follows the following
schema:

```yaml
"$id": https://pipeline.mlx.org/output.schema.json"
"$schema": http://json-schema.org/draft-07/schema#"
type: object
title: Output
properties:
  error_status: { "$ref": "#/definitions/OutputErrorStatus" }
  outputs:
    type: object
  exec_properties:
    type: object
definitions:
  OutputErrorStatus:
    type: object
    properties:
      code:
        type: string
        enum: [PERMANENT_ERROR, RETRYABLE_ERROR]
      message:
        type: string
```

> The output.json file is optional, but if user’s container writes to the file. It

Reviewer 3: ```suggestion
The output.json file is optional, but if the user’s container writes to the file. It
```

> overrides the default handling of k8s pod launcher. Here are the explanation of

Reviewer 3: ```suggestion
overrides the default handling of the k8s pod launcher. The output fields are:
```

> the output fields:

Reviewer 3: ```suggestion
```

> *   error_status: it tells the executor whether it should retry or fail.

Reviewer 3: ```suggestion
*   error_status: tells the executor whether it should retry or fail
```

> *   outputs and exec_properties: they will be used to override the execution and

Reviewer 3: ```suggestion
*   outputs and exec_properties: used to override the execution and
```

    output artifact metadata in MLMD.

> The output interfaces relies on `BaseComponentLauncher` to update states back to

Reviewer 3: ```suggestion
The output interfaces rely on `BaseComponentLauncher` to update states back to
```

MLMD from executor.

### Auth context resolution

> K8s pod launcher internally uses k8s python client. The auth context resolution

Reviewer 3: ```suggestion
The k8s pod launcher internally uses the k8s Python client. The auth context resolution
```

logic is as follows:

> 1.  If the current env is in a cluster, uses `load_incluster_config` to load k8s

Reviewer 3: ```suggestion
1.  If the current env is in a cluster, use `load_incluster_config` to load k8s
```

    context.
1.  If not, use default k8s active context to connect to remote cluster.

### Pod launcher resiliency

In this design section, we focused more on the launcher resiliency under
`KubeflowDAGRunner`. In `AirflowDAGRunner`, the launcher code is running in the

> same process of Airflow orchestrator which we rely on Airflow to ensure its

Reviewer 3: ```
same process of Airflow orchestrator which we rely on Airflow to ensure its
```

in which? I'm not sure what this sentence means: there is a word (or two) missing.

Author: Reword the sentence to make it more clear.

resiliency. `BeamDAGRunner`, however, is considered mainly for local testing

> purpose and we won't add support for it to be resiliency.

Reviewer 3: ```suggestion
purpose and we won't add support for it to be resilient.
```

In `KubeflowDAGRunner`, a pipeline step will create two pods in order to execute
user’s container:

> *   A launcher pod which contains driver, k8s pod launcher and publisher code.

Reviewer 3: ```suggestion
*   A launcher pod which contains the driver, k8s pod launcher, and publisher code.
```

*   A user pod with user’s container.

> Pod in k8s is not resilient by itself. We will use Argo’s retry feature to make

Reviewer 3: ```suggestion
A pod in k8s is not resilient by itself. We will use Argo’s retry feature to make
```

> launcher pod to be partially resilient. The details are as follows:

Reviewer 3: ```suggestion
the launcher pod partially resilient. The details are as follows:
```

> *   Each argo launcher step will be configured with a default retry count.

Reviewer 3: ```suggestion
*   Each Argo launcher step will be configured with a default retry count.
```

> *   Argo will retry the step in case of failure no matter what type of error.

Reviewer 3: ```suggestion
*   Argo will retry the step in case of failure, no matter what type of error.
```

> *   The launcher container will create a tmp workdir from `pipeline_root`.

Reviewer 3: ```suggestion
*   The launcher container will create a tmp workdir in `pipeline_root`.
```
?

> *   It will keep intermediate results like created pod ID in the tmp workdir.

Reviewer 3: ```suggestion
*   It will keep intermediate results (for example, the ID of the created pod) in the tmp workdir.
```

*   The k8s pod launcher will be implemented in a way that it will resume the
    operation based on the intermediate results in the tmp workdir.
*   The launcher will also record a permanent failure data in the tmp workdir so
    it won’t resume the operation in case of non-retriable failures.

### Default retry strategy

K8s pod launcher supports exponential backoff retry. This strategy applies to

> all runners which can support k8s pod launcher. Docker launcher are not in the

Reviewer 3: ```suggestion
all runners which can support k8s pod launcher. Docker launchers are not in the
```

scope of the design as it is mainly for local development use case.

The retry only happens if the error is retriable. An error is retriable only
when:

> 

Reviewer 3: AND or OR?
(All three required, or any?)

Author: It's OR. I added it explicitly.

*   It’s a transient error code from k8s pod API.
*   The output.json file from artifact store indicates it’s a retriable error.
*   The pod get deleted (For example: GKE preemptible pod feature).

### Log streaming

> Container launcher streams log from user’s docker container or k8s pod through

Reviewer 3: ```suggestion
The container launcher streams the log from user’s docker container or k8s pod through the
```

Reviewer 2: _user’s Docker container_  - caps for consistency

> API. It will start a thread which constantly pulls new logs and output it to

Reviewer 3: ```suggestion
API. It will start a thread which constantly pulls new logs and outputs them to
```

local stdout.

### Cancellation

> The container launcher handles cancellation request varies by orchestrators:

Reviewer 3: ```suggestion
How the container launcher handles cancellation request varies by orchestrators:
```

*   Airflow natively supports cancellation propagation to operator. We will need
    to pass the cancellation request from operator into executor.
*   Argo doesn’t natively support cancellation propagation. Currently KFP relies

>     on Argo’s pod annotation to workaround the limitation and have been proven

Reviewer 3: ```suggestion
    on Argo’s pod annotation to workaround the limitation and has been proven
```

>     to be working. We will use the same way to propagate cancellation request to

Reviewer 3: ```suggestion
    to work. We will use the same process to propagate cancellation requests to
```

    user’s container.

> In order to allow user to specify the cancellation command line entrypoint, k8s

Reviewer 3: ```suggestion
In order to allow the user to specify the cancellation command line entrypoint, the k8s
```

pod launcher will support an optional parameter called `cancellation_command`
from `ExecutorContainerSpec`.

## Open discussions

> *   In argo runner, each step requires 2 pods with total 3 containers (launcher

Reviewer 3: ```suggestion
*   In the Argo runner, each step requires 2 pods with total 3 containers (launcher
```

    main container + launcher argo wait container + user main container) to run.

>     Although all launcher containers requires minimum k8s resources, it still

Reviewer 3: ```suggestion
    Although each launcher container requires minimal k8s resources,
```

>     can be a concern on resource usage.

Reviewer 3: ```suggestion
    resource usage is still a concern.
```
